### DISCUSSION TITLE
discuss: how round payout happens

### DISCUSSION THREAD

**Discussion by thelostone-mc** [dcgtc/dgrants#55] [Architecture & Design] [answered]
##### Description 

Right now on `GrantRound` once the round end there is a method `payoutGrants` which the `GrantRound` owner invokes to transfer funds to another address  
https://github.com/dcgtc/dgrants/blob/main/contracts/contracts/GrantRound.sol#L103-L107

How does this flow work post that?

This is my understanding as it stands.  

1. Round Ends.
2. Anyone can calculate distribution using the clr match algorithm (mostly done by the user who created the GrantRound ) 
3. They deploy `MerkleDistributor` contract with :
   - merkle root  : to make it easier for others to validate the distribution 
   - actual distribution 
4. Once the contract is deployed -> there will be an action by `approval_address` (currently gitcoin but later on DAO )  to validate and approve the distribution 
5. The Round creator invokes funds `payoutGrants` and transfer funds to the approved `MerkleDistributor` contract
6. once the funds are moved. we have to pick one of  2 options
   a) Grant owner claim funds from this contract (this will have to be tracked on who claimed and who didn't)
   b) the address that deployed the `MerkleDistributor` does a bulk transfer and transfer on the contract

Q: 
- Do we need the `MerkleDistributor` contract, can this not be stored on the GrantRound contract ? (Are we avoiding it cause it would be expensive ? -> we are already storing the list of grants)
- In step 4 -> how do we decide what the approval_address is

**Comment by apbendi** [comment, marked as answer]
Hey @thelostone-mc, separating the payout contract from the Round contract is a matter of separation of concerns. We want  the Round contract managing the minimal, core set of functions which define a Grants Round.

To achieve this, we use an `address` as an abstraction barrier. The Round doesn't know what's on the other side of the address. It could be an EOA, or it could be a contract with any arbitrary logic.

We actually leverage this in two places here: 1) the privileged address that can transfer funds to a payout address after the Round is over and 2) the address where the payout actually goes. By simply using an address in each case, the Round contract remains simple and easy to test. It also means the concept of a Round can be more flexible, which is good both for other entities who might want to run Rounds and make different sets of tradeoffs, and for us as well, as we'll have more flexibility to iterate on more advanced approaches over time. Does this make sense as a general concept?

In the specific case of Gitcoin dGrants MVP, we've discussed pursuing something like this:

1) The privileged address that would have the right to transfer the payout funds would be the DAO (specifically, the DAO's timelock contract, which can execute any arbitrary batch of transactions that pass as a proposal)
2) The payout contract itself would either be:
    * A naive mapping distributor, such as the one used by cGrants currently, where the recipients and amounts are all stored on chain
    * A slightly more advanced MerkleDistributor contract, such that only a Merkle root would need to be stored on chain, and a Merkle proof would be submitted when claiming funds

There are specific tradeoffs to each of the options for payout contract, along with sub-decisions to be made in each case, but the important point here is that **the Round contract shouldn't know anything about these**.

In either case, the high level workflow should be something like this:

1. The round ends
2. We run the matching script to generate payouts
3. We deploy an immutable contract of some kind which codifies those payouts
4. We submit a proposal to the DAO which transfers the matching funds to the payout contract
5. The DAO checks our work and votes to approve or reject the approval

Note that in the description above, "We" could *technically* be anyone. Only the DAO has authority to transfer the funds, and thus who does steps 2-4 is irrelevant, as long as the DAO members verify it independently.

Does all of this make sense?
